File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    char[][] board;
    int numRows;
    int numCols;


    public void solve(char[][] board) {
        this.board = board;
        numRows = board.length;
        numCols = board[0].length;

        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numCols; j++) {
                if ((i == 0 || i == numRows - 1 ||
                    j == 0 || j == numCols - 1) &&
                    board[i][j] == 'O') {
                    dfs(i, j);
                }
            }
        }

        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numCols; j++) {
                if (board[i][j] == '.') {
                    board[i][j] = 'O';
                } 
                else if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }
    }

    private void dfs(int row, int col) {
        if (row < 0 || row == numRows ||
            col < 0 || col == numCols)
            return;
        if (board[row][col] != 'O') 
            return;

        board[row][col] = '.';

        dfs(row + 1, col);
        dfs(row - 1, col);
        dfs(row, col + 1);
        dfs(row, col - 1);
    }

}

File: 0130-surrounded-regions/0130-surrounded-regions.py
from typing import List

class Solution:
    def __init__(self):
        self.board: List[List[str]] = []
        self.num_rows = 0
        self.num_cols = 0

    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        if not board or not board[0]:
            return

        self.board = board
        self.num_rows = len(board)
        self.num_cols = len(board[0])

        # Mark border-connected 'O's
        for i in range(self.num_rows):
            for j in range(self.num_cols):
                if (i == 0 or i == self.num_rows - 1 or
                    j == 0 or j == self.num_cols - 1) and self.board[i][j] == 'O':
                    self.dfs(i, j)

        # Flip the cells accordingly
        for i in range(self.num_rows):
            for j in range(self.num_cols):
                if self.board[i][j] == '.':
                    self.board[i][j] = 'O'
                elif self.board[i][j] == 'O':
                    self.board[i][j] = 'X'

    def dfs(self, row: int, col: int) -> None:
        if (row < 0 or row >= self.num_rows or
            col < 0 or col >= self.num_cols or
            self.board[row][col] != 'O'):
            return

        self.board[row][col] = '.'

        self.dfs(row + 1, col)
        self.dfs(row - 1, col)
        self.dfs(row, col + 1)
        self.dfs(row, col - 1)

        